Add Output SpinState to SampleLog in PolarizationCorrectionFredrikze Algorithm

yusufjimoh opened the pull request:
### Description of work
- Added functions to include spin state information in the PolarizationCorrectionFredrikZe algorithm.
- Modified mapOutputWorkspaceToSpinState to append output spin states to the sample log when requested.
- Updated unit tests to verify that the order of output workspaces aligns with the specified output spin states and that spin state information is correctly recorded in the sample log.
- Updated documentation to reflect the new functionality and clarify its usage

Fixes #38265 

### To test:
## Test for PA
1) Run the following script. It will output a group of workspaces called "out". Check that the sample log of each child workspace **does not** contain an entry with the name `spin_state_ORSO`.
```
run1 = Load("POLREF00004699.nxs")
run2 = run1 * 1.2
run = GroupWorkspaces([run1, run2])
ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, run.getItem(0), True)

out = PolarizationCorrectionFredrikze(run, eff, "PA", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="pp, pa, ap, aa")
```

2) Without clearing any of your workspaces, run the following:

```
with_log = PolarizationCorrectionFredrikze(run, eff, "PA", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="pp, pa, ap, aa", AddSpinStateToLog=True)
```
Check that the sample log of each child workspace in the "with_log" group **does** contain an entry with the name `spin_state_ORSO`. The log values should be as follows:

- `with_log_1` workspace has value `pp`
- `with_log_2` workspace has value `pm`
- `with_log_3` workspace has value `mp`
- `with_log_4` workspace has value `mm`

3) Test the functionality on the GUI using the same input workspaces and spinstates
- when AddSpinStateToLog checkbox is not enabled: Check that the sample log of each child workspace **does not** contain an entry with the name `spin_state_ORSO`.
- 
- when addSpinStateToLog checkbox is enabled: Check that the sample log of each child workspace in the "with_log" group **does** contain an entry with the name `spin_state_ORSO`. The log values should be as follows:

- `with_log_1` workspace has value `pp`
- `with_log_2` workspace has value `pm`
- `with_log_3` workspace has value `mp`
- `with_log_4` workspace has value `mm`

## Test for PNR
1) clear your workspaces and run the following script. It will output a group of workspaces called "out". Check that the sample log of each child workspace **does not** contain an entry with the name `spin_state_ORSO`.
```
run = Load("POLREF00004699.nxs")
ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, run.getItem(0), True)

out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="p, a", OutputSpinStates="p, a")
```

2) Without clearing any of your workspaces, run the following:

```
with_log = out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="p, a", OutputSpinStates="p, a", AddSpinStateToLog=True)
```
Check that the sample log of each child workspace in the "with_log" group **does** contain an entry with the name `spin_state_ORSO`. The log values should be as follows:

- `with_log_1` workspace has value `po`
- `with_log_2` workspace has value `mo`

3) Test the functionality on the GUI using the same input workspaces and spinstates
- when AddSpinStateToLog checkbox is not enabled: Check that the sample log of each child workspace **does not** contain an entry with the name `spin_state_ORSO`.

- when addSpinStateToLog checkbox is enabled: Check that the sample log of each child workspace in the "with_log" group **does** contain an entry with the name `spin_state_ORSO`. The log values should be as follows:

- `with_log_1` workspace has value `po`
- `with_log_2` workspace has value `mo`

yusufjimoh committed:
Add Output Spinstate to samplelog in PolarizationCorrection Fredrikze Algorithm

- added function to add spinstate to polarizationcorrection fredrikze algorithm
- modified mapoutputworkspacetospinstate to add outspinstate to samplelog if requested
- updated unit tests to check that the order of output workspaces matches the order of specified output spinstates in the samplelog
- updated documentation

yusufjimoh committed:
Remove unused OutputWorkspaceName parameter  from execPA and execPNR

adriazalvarez reviewed (changes requested):
Thanks for adding this feature to the correction algorithm. I have tested it with the example scripts and can confirm that the `spin_state_ORSO` log is correctly displayed when the `AddSpinStateToLog` is checked to true.
Thanks also for adding an extensive test coverage, I have added a few suggestions to improve readability of the tests, and I'm not sure whether we should separate the log checking for complete different tests instead of appending them to existing ones, but for the time being, see if you can work on these suggestions.

yusufjimoh committed:
Update Test based on suggested changes from code
- remove redundant unit tests
- limited the use of hardcoded strings especially for strings used in multiple test functions

yusufjimoh committed:
refactored test to reduce code repetition

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -25,7 +26,14 @@ using namespace Mantid::API;
 using namespace WorkspaceCreationHelper;
 
+namespace {
+
+static const int PA_GROUP_SIZE = 4;
+static const int PNR_GROUP_SIZE = 2;
```
```suggestion
constexpr int PA_GROUP_SIZE = 4;
constexpr int PNR_GROUP_SIZE = 2;
```

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -305,20 +327,38 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+    for (size_t orderIdx = 0; orderIdx < spinStateOrders.size(); ++orderIdx) {
+      alg->setProperty("InputSpinStates", spinStateOrders[orderIdx]);
+      alg->setProperty("OutputSpinStates", spinStateOrders[orderIdx]);
+      alg->setProperty("AddSpinStateToLog", true);
```
```suggestion
```

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -336,13 +376,53 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+    for (size_t orderIdx = 0; orderIdx < spinStateOrders.size(); ++orderIdx) {
+      alg->setProperty("InputSpinStates", spinStateOrders[orderIdx]);
+      alg->setProperty("OutputSpinStates", spinStateOrders[orderIdx]);
+      alg->setProperty("AddSpinStateToLog", true);
```
```suggestion
```

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -336,13 +376,53 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+
+  void test_valid_spin_state_order_for_PNR() {
+    auto groupWS = createGroupWorkspace(2, 4, 1, 1);
+    auto eff = makeEfficiencies(create1DWorkspace(4, 1, 1), "1,0,0,0", "1,0,0,0");
```
```suggestion
```

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -336,13 +376,53 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+    auto groupWS = createGroupWorkspace(2, 4, 1, 1);
+    auto eff = makeEfficiencies(create1DWorkspace(4, 1, 1), "1,0,0,0", "1,0,0,0");
+    auto alg = initializeAlgorithm(groupWS, "PNR", "1,0,0,0");
+
```
```suggestion
alg->setProperty("AddSpinStateToLog", true);

```

adriazalvarez reviewed (changes requested):
Few more small changes

rbauststfc commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -305,20 +327,38 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+  void test_valid_spin_state_order_for_PA() {
     auto groupWS = createGroupWorkspace(4, 4, 1, 1);
     auto alg = initializeAlgorithm(groupWS, "PA", "1,0,0,0");
 
```
Hi Yusuf, just to clarify following our conversation, Adrian is suggesting here that you would have the following:

```
  void test_valid_spin_state_order_for_PA() {
    auto groupWS = createGroupWorkspace(4, 4, 1, 1);
    auto alg = initializeAlgorithm(groupWS, "PA", "1,0,0,0");
    alg->setProperty("AddSpinStateToLog", true);
```

This way we're setting the `AddSpinStateToLog` property to `true`, which this test requires, but we're doing it outside the for loop that we have later in the test. This approach should be more efficient because this property will have the same value for every iteration of the loop, so we don't need to keep setting it each time.

As we discussed, you could potentially add something to `initializeAlgorithm` to optionally set this property when required. However I think most of the tests in this class do not need to set this property, so it seems fine to leave it to be set separately. If you prefer to move it into `initializeAlgorithm`, though, then that would also be fine I think.

Commit:
implemented suggested changes from pull request
-removed addSpinStateToLog porpery from within for loop
- change constants declared with static int to constexpr
- remove unused variables in test methods

adriazalvarez reviewed (changes requested):
Thanks for the latest changes,I have spotted couple of tiny changes.

adriazalvarez reviewed (dismissed):
Thanks for the changes.

rbauststfc reviewed (dismissed):
Before I gatekeeper this one, I've just noticed a couple of local variables in the tests that don't look like they're following our C++ naming conventions. I've also suggested a slight tweak to the documentation update.

yusufjimoh committed:
change some variable names to match mantid naming cinvention

rbauststfc commented:
Thanks @yusufjimoh for making the requested changes. @adriazalvarez, if you're happy to re-approve then I'll get this merged in.
